docs/proposals/002-api-proposal/proposal.md

```diff
@@ -0,0 +1,313 @@
+        Weight int
+}
+
+// Objective captures the latency SLO of a LLM use case.
+// In MVP, meeting the SLO is on a best effort basis.
+// Future: Extend the API for different behaviors of meeting the SLO.
+// The gateway will perform best-effort load balancing, and work with other components (e.g., autoscaler) to meet the
+// objectives. 
+type Objective struct {
+        // The AverageLatencyPerOutputToken is calculated as the e2e request latency divided by output token 
+        // length. Note that this is different from what is known as TPOT (time per output token) which only 
+        // takes decode time into account.
+        // The P95 is calculated over a fixed time window defined at the operator level.
+        DesiredAveragePerOutputTokenLatencyAtP95OverMultipleRequests
+    *time.Duration
+}
+```
+
+### Yaml Examples
+
+#### BackendPool(s)
+Here we create 2 BEPs that subscribe to services to collect the appropriate pods
+```yaml
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: BackendPool
+metadata:
+  name: llama-2-pool
+  services: 
+  - llama-2-vllm
+---
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: BackendPool
+metadata:
+  name: gemini-pool
+  services: 
+  - gemini-jetstream-tpu-v5e
+  - gemini-vllm-a100
+```
+
+#### Model Group
+
+Here we consume both pools with a single ModelGroup, while also specifying 2 useCases. Where `sql-code-assist` is both the name of the ModelUseCase, and the name of the LoRA adapter on the model server. And `npc-bot` has a layer of indirection for those names, as well as a specified objective.
+```yaml
+apiVersion: inference.x-k8s.io/v1alpha1
+kind: ModelGroup
+metadata:
+  name: my-model-group
+spec:
+  useCases:
+  - modelName: sql-code-assist
+  - modelName: npc-bot
+    objective: 
+      desiredAveragePerOutputTokenLatencyAtP95OverMultipleRequests: 50ms
+    targetModels:
+      targetModelName: npc-bot-v1
+        weight: 50
+      targetModelName: npc-bot-v2
+        weight: 50 	
+  poolRef: 
+   - name: llama-2-pool
+   - name: gemini-pool
+```
+
+### Diagrams
+
+Much of this is better explained visually:
+
+Below is a detailed view of the BackendPool
+
+![BackendPool](./images/bep.svg)
+
+This diagram lightly follows the example request for a model `interestingName`. 
+The flow can be described as:
+- The request comes in to our routing solution(Ext-Proc)
+- ExtProc looks up the UseCases affiliated with this pool `examplePool`
+- `interestingName` is currently undergoing a change of LoRA adapters from `creativeNameGen-v3` (20% traffic split) to `veryCreativeNameGen` (80% traffic split)
+- `veryCreativeNameGen` is selected as the LoRA adapter, and replaces `interestingName` in the body of the request (mutated by ext-proc) 
+- the request is then efficiently scheduled onto one of the valid Pods
+- metrics are sent back to the BEP, aggregated and re-emitted via sidecar (following the metric standardization)
```

Comment:
How about highlight the format of metrics here? Like in Prometheus format or OTEL etc? I think this can help the end user to integrate with some observability platforms

Reply:
Done! 

Agreed that our metrics story could use some improvement. We intend to create a protocol of metrics that the I-GW expects to ingest. While not the scope of this section here, would also be equally valuable that we specify any metrics we intend to emit. I don't expect this portion to be solidified for the alpha phase however.